Source/Core/Common/MemoryUtil.cpp

```diff
@@ -79,25 +146,24 @@ void* AllocateExecutableMemory(size_t size, bool low)
   {
     ptr = nullptr;
 #endif
-    PanicAlert("Failed to allocate executable memory. If you are running Dolphin in Valgrind, try "
-               "'#undef MAP_32BIT'.");
+    PanicAlert("Failed to allocate executable memory.");
   }
-#if !defined(_WIN32) && defined(_M_X86_64) && !defined(MAP_32BIT)
+#if !defined(_WIN32) && defined(_M_X86_64)
   else
   {
-    if (low)
+    if (map_hint <= 0xFFFFFFFF)
     {
-      map_hint += size;
-      map_hint = (char*)RoundPage((uintptr_t)map_hint); /* round up to the next page */
+      // Round up if we're below 32-bit mark, probably allocating sequentially.
+      map_hint += RoundPage(size);
+
+      // If we moved ahead too far, skip backwards and recalculate.
+      // When we free, we keep moving forward and eventually move too far.
+      if (map_hint - hint_location >= 0x70000000)
```

Comment:
I feel like this `0x70000000` should be named (or at least explained) too, especially since it doesn't look like any familiar kind of mask/limit/whatever (like, as yourself: why is it `0x70000000` specifically and not `0x60000000`, `0x80000000` or any other number).